CONFIGS.md
# Configurations Documentation 

1. [Experiment Parameter Definitions](#1-experiment-parameters)
2. [MCTS Parameters](#2-mcts-specific-defaults-parameters)
3. [DQN Parameters](#3-dqn-specific-defaults-parameters)
4. [Example MCTS config](#example-mcts-config-file)
5. [Example DQN config](#example-dqn-config-file)

___


All configuration files (.yaml) must contain `[Methods]` and `[Defaults]` sections.  
The `[Methods]` section contains general experiment settings.   
The `[Defaults]` section contains settings specific to the motion planning method. 
``` yaml
[Methods]
...

[Defaults]
...
```
## 1. Experiment Parameters 
``` yaml
[Methods]
method : string
    Choose from ['mcts','dqn']
sensor : string
    Choose from ['drone','signalstrength', 'bearing']
reward : string
    Choose from ['range_reward', 'entropy_collision_reward']
target_speed : float
    Speed of the target
target_start : float
    Starting distance between target and sensor
resample_proportion: float
    Ratio of particles to resample from the prior at every particle filter update
action : simpleactions
state : rfstate 
```

## 2. MCTS specific `[Defaults]` Parameters 
``` yaml
[Defaults]
simulations : int
    Number of simulations
depth : int
    Tree depth
lambda_arg : float
    Lambda value
trials : int
    Number of runs
iterations : int
    Number of iterations
collision : float
    Reward value for collision
loss : float
    Reward value for loss function
plotting : bool
    Flag to plot or not
```

## 3. DQN specific `[Defualts]` parameters 
``` yaml
[Defaults]
log_path : string
    Path for logging output
use_gpu : bool
    If True use GPU device
number_timesteps : int
    Number of timesteps
dueling : bool
    If True dueling value estimation will be used
save_path : string
    Path for saving
save_interval : int
    Interval for saving output values
ob_scale : int
    Scale for observation
gamma : float
    Gamma input value
grad_norm : float
    Max norm value of the gradients to be used in gradient clipping
double_q  : bool
    If True double DQN will be used
param_noise : bool
    If True use parameter space noise
exploration_fraction : float
    Fraction of entire training period over which the exploration rate is annealed
exploration_final_eps : float
    Final value of random action probability
batch_size : int
    Size of a batched sampled from replay buffer for training
train_freq : int
    Update the model every `train_freq` steps
learning_starts : int
    How many steps of the model to collect transitions for before learning starts
target_network_update_freq : int
    Update the target network every `target_network_update_freq` steps
buffer_size : int
    Size of the replay buffer
prioritized_replay : bool
    If True prioritized replay buffer will be used.
prioritized_replay_alpha : float
    Alpha parameter for prioritized replay
prioritized_replay_beta0 : float
    Beta parameter for prioritized replay
atom_num : int
    Atom number in distributional RL for atom_num > 1
min_value : float
    Min value in distributional RL
max_value : float
    Max value in distributional RL
plotting : bool
    Flag to plot or not
```

## 4. Example [MCTS config file](configs/mcts.yaml)
``` yaml
[Methods]
method : mcts
action : simpleactions
sensor : drone
state : rfstate 
target_speed: 1.0
target_start: 100
resample_proportion: 0.005
reward: range_reward

[Defaults]
lambda_arg : 0.8
collision : -2.
loss : -2.
depth : 10
simulations : 200
plotting : False
trials : 500
iterations : 150
```

## 5. Example [DQN config file](configs/dqn.yaml) 
``` yaml
[Methods]
method : dqn
action : simpleactions
sensor : drone
state : rfstate 
target_speed: 1.0
target_start: 100
resample_proportion: 0.1
reward: entropy_collision_reward

[Defaults]
number_timesteps : 100000
dueling : True
double_q : True
param_noise : True
exploration_fraction : 0.2
exploration_final_eps : 0.1
batch_size : 64
train_freq : 4
learning_starts : 100
target_network_update_freq : 600
buffer_size : 10000
prioritized_replay : True
prioritized_replay_alpha : 0.6
prioritized_replay_beta0 : 0.4
min_value : -10
max_value : 10
max_episode_length : 150
atom_num : 1
ob_scale : 1
gamma : 0.99
grad_norm : 10.0
save_interval : 100000
save_path : checkpoints
log_path : rl_log
use_gpu : True
plotting : False
```

docs/CONFIGS.md
# Configurations Documentation 

1. [Experiment Parameter Definitions](#1-experiment-parameters)
2. [MCTS Parameters](#2-mcts-specific-defaults-parameters)
3. [DQN Parameters](#3-dqn-specific-defaults-parameters)
4. [Example MCTS config](#4-example-mcts-config-file)
5. [Example DQN config](#5-example-dqn-config-file)

___


All configuration files (.yaml) must contain `[Methods]` and `[Defaults]` sections.  
The `[Methods]` section contains general experiment settings.   
The `[Defaults]` section contains settings specific to the motion planning method. 
``` yaml
[Methods]
...

[Defaults]
...
```
## 1. Experiment Parameters 
``` yaml
[Methods]
method : string
    Choose from ['mcts','dqn']
sensor : string
    Choose from ['drone','signalstrength', 'bearing']
reward : string
    Choose from ['range_reward', 'entropy_collision_reward']
target_speed : float
    Speed of the target
target_start : float
    Starting distance between target and sensor
resample_proportion: float
    Ratio of particles to resample from the prior at every particle filter update
action : simpleactions
state : rfstate 
```

## 2. MCTS specific `[Defaults]` Parameters 
``` yaml
[Defaults]
simulations : int
    Number of simulations
depth : int
    Tree depth
lambda_arg : float
    Lambda value
trials : int
    Number of runs
iterations : int
    Number of iterations
collision : float
    Reward value for collision
loss : float
    Reward value for loss function
plotting : bool
    Flag to plot or not
```

## 3. DQN specific `[Defaults]` parameters 
``` yaml
[Defaults]
log_path : string
    Path for logging output
use_gpu : bool
    If True use GPU device
number_timesteps : int
    Number of timesteps
dueling : bool
    If True dueling value estimation will be used
save_path : string
    Path for saving
save_interval : int
    Interval for saving output values
ob_scale : int
    Scale for observation
gamma : float
    Gamma input value
grad_norm : float
    Max norm value of the gradients to be used in gradient clipping
double_q  : bool
    If True double DQN will be used
param_noise : bool
    If True use parameter space noise
exploration_fraction : float
    Fraction of entire training period over which the exploration rate is annealed
exploration_final_eps : float
    Final value of random action probability
batch_size : int
    Size of a batched sampled from replay buffer for training
train_freq : int
    Update the model every `train_freq` steps
learning_starts : int
    How many steps of the model to collect transitions for before learning starts
target_network_update_freq : int
    Update the target network every `target_network_update_freq` steps
buffer_size : int
    Size of the replay buffer
prioritized_replay : bool
    If True prioritized replay buffer will be used.
prioritized_replay_alpha : float
    Alpha parameter for prioritized replay
prioritized_replay_beta0 : float
    Beta parameter for prioritized replay
atom_num : int
    Atom number in distributional RL for atom_num > 1
min_value : float
    Min value in distributional RL
max_value : float
    Max value in distributional RL
plotting : bool
    Flag to plot or not
```

## 4. Example [MCTS config file](configs/mcts.yaml)
``` yaml
[Methods]
method : mcts
action : simpleactions
sensor : drone
state : rfstate 
target_speed: 1.0
target_start: 100
resample_proportion: 0.005
reward: range_reward

[Defaults]
lambda_arg : 0.8
collision : -2.
loss : -2.
depth : 10
simulations : 200
plotting : False
trials : 500
iterations : 150
```

## 5. Example [DQN config file](configs/dqn.yaml) 
``` yaml
[Methods]
method : dqn
action : simpleactions
sensor : drone
state : rfstate 
target_speed: 1.0
target_start: 100
resample_proportion: 0.1
reward: entropy_collision_reward

[Defaults]
number_timesteps : 100000
dueling : True
double_q : True
param_noise : True
exploration_fraction : 0.2
exploration_final_eps : 0.1
batch_size : 64
train_freq : 4
learning_starts : 100
target_network_update_freq : 600
buffer_size : 10000
prioritized_replay : True
prioritized_replay_alpha : 0.6
prioritized_replay_beta0 : 0.4
min_value : -10
max_value : 10
max_episode_length : 150
atom_num : 1
ob_scale : 1
gamma : 0.99
grad_norm : 10.0
save_interval : 100000
save_path : checkpoints
log_path : rl_log
use_gpu : True
plotting : False
```

birdseye/env.py
import numpy as np
import random
from .utils import pol2cart, particles_mean_belief
from birdseye.pfrnn.pfrnn import pfrnn
from pfilter import ParticleFilter, systematic_resample


class RFEnv(object):

    def __init__(self, sensor=None, actions=None, state=None, simulated=False):
        # Sensor definitions
        self.sensor = sensor
        # Action space and function to convert from action to index and vice versa
        self.actions = actions
        # Setup initial state
        self.state = state
        # Flag for simulation vs real data
        self.simulated = simulated

        self.pfrnn = pfrnn()

    def dynamics(self, particles, control=None, **kwargs):
        """Helper function for particle filter dynamics

        Returns
        -------
        array_like
            Updated particle state information
        """
        return np.array([list(self.state.update_state(p, control)) for p in particles])

    def reset(self, num_particles=2000):
        """Reset initial state and particle filter

        Parameters
        ----------
        num_particles : integer
            Number of particles to build particle filter

        Returns
        -------
        env_obs : array_like
            Heatmap distribution of observed particles from reset filter
        """

        self.iters = 0
        self.state.target_state = self.state.init_target_state()
        self.state.sensor_state = self.state.init_sensor_state()

        # Setup particle filter
        self.pf = ParticleFilter(
                        prior_fn=lambda n: np.array([self.state.random_state() for i in range(n)]),
                        observe_fn=lambda states, **kwargs: np.array([np.array(self.sensor.observation(x)) for x in states]),
                        n_particles=num_particles,
                        dynamics_fn=self.dynamics,
                        noise_fn=lambda x, **kwargs: x,
                        resample_proportion=0.005,
                        #noise_fn=lambda x:
                        #            gaussian_noise(x, sigmas=[0.2, 0.2, 0.1, 0.05, 0.05]),
                        weight_fn=lambda hyp, o, xp=None,**kwargs: [self.sensor.weight(None, o, state=x) for x in xp],
                        resample_fn=systematic_resample,
                        column_names = ['range', 'bearing', 'relative_course', 'own_speed'])

        env_obs = self.env_observation()
        return env_obs


    # returns observation, reward, done, info
    def step(self, action_idx):
        """Function to make step based on
           state variables and action index

        Parameters
        ----------
        action_idx : integer
            Index for action to make step

        Returns
        -------
        env_obs : array_like
            Heatmap distribution of observed particles from filter
        reward : float
            Reward value for specified action
        0 : int
            Placeholder integer value
        info : dict
            Dictionary to track step specific values (reward, iteration)
        """

        # Get action based on index
        action = self.actions.index_to_action(action_idx)
        # Determine next state based on action & current state variables
        next_state = self.state.update_state(self.state.target_state, action)
        # Update absolute position of sensor
        self.state.update_sensor(action)
        # Get sensor observation
        observation = self.sensor.observation(next_state)
        # Update particle filter
        self.pf.update(np.array(observation), xp=self.pf.particles, control=action)
        # Calculate reward based on updated state & action
        reward = self.state.reward_func(state=next_state, action_idx=action_idx, particles=self.pf.particles)
        #reward = -1. * self.get_distance_error()
        # Update the state variables
        self.state.target_state = next_state

        env_obs = self.env_observation()
        self.iters += 1
        info = {'episode':{}}
        info['episode']['l'] = self.iters
        info['episode']['r'] = reward
        info['observation'] = observation

        return (env_obs, reward, 0, info)

    def entropy_collision_reward(self, state, action_idx=None, delta=10, collision_weight=1):
        pf_r = self.pf.particles[:,0]
        pf_theta = np.radians(self.pf.particles[:,1])
        pf_x, pf_y = pol2cart(pf_r, pf_theta)
        xedges = np.arange(-150, 153, 3)
        yedges = np.arange(-150, 153, 3)
        b = np.histogram2d(pf_x, pf_y, bins=(xedges, yedges))
        b /= np.sum(b)
        b += 0.0000001

        H = -1. * np.sum([b * np.log(b)])
        collision_rate = np.mean(self.pf.particles[:,0] < delta)
        cost = H + collision_weight * collision_rate

        return -1. * cost

    def env_observation(self):
        """Helper function for environment observation

        Returns
        -------
        array_like
            Heatmap distribution of current observed particles
        """
        #return np.expand_dims(self.particle_heatmap_obs(self.pf.particles), axis=0)
        pf_map = np.expand_dims(self.particle_heatmap_obs(self.pf.particles), axis=0).reshape(-1)
        _,_,_,_, mean_r, mean_theta, mean_heading, mean_spd = particles_mean_belief(self.pf.particles)
        mean_belief = np.array([mean_r, mean_theta, mean_heading, mean_spd])
        return np.concatenate((mean_belief, pf_map))

    def particle_heatmap_obs(self, belief):
        """Function to build histogram representing
           belief distribution in cart coords

        Parameters
        ----------
        belief : array_like
            Belief distribution parameters

        Returns
        -------
        heatmap : array_like
            Histogram of belief state
        """
        # Transformation of belief to cartesian coords
        cart  = np.array(list(map(pol2cart, belief[:,0], np.radians(belief[:,1]))))
        x = cart[:,0]
        y = cart[:,1]

        # Build two-dim histogram distribution
        xedges = np.arange(-150, 153, 3)
        yedges = np.arange(-150, 153, 3)
        heatmap, xedges, yedges = np.histogram2d(x, y, bins=(xedges, yedges))

        return heatmap

    def get_absolute_particles(self):
        return np.array([self.state.get_absolute_state(t) for t in self.pf.particles])

    def get_absolute_target(self):
        return self.state.get_absolute_state(self.state.target_state)

    def get_particle_centroid(self):

        particles = self.pf.particles

        particles_r = particles[:,0]
        particles_theta = np.radians(particles[:,1])
        particles_heading = particles[:,2]
        particles_x, particles_y = pol2cart(particles_r, particles_theta)

        # centroid of particles x,y
        mean_x = np.mean(particles_x)
        mean_y = np.mean(particles_y)

        return mean_x, mean_y

    def get_distance_error(self):

        mean_x, mean_y = self.get_particle_centroid()

        target_r = self.state.target_state[0]
        target_theta = np.radians(self.state.target_state[1])
        target_x, target_y = pol2cart(target_r, target_theta)

        centroid_distance_error = np.sqrt((mean_x - target_x)**2 + (mean_y - target_y)**2)

        return centroid_distance_error
